### PR TITLE
IRV-372 Implement a Context API (data provider) for templates.

### PR DESCRIPTION
**Pull request by joemcgill**
The main thing this does is add a new context API so components can declare that they both provide or use context properties during component registration by adding a `provides_context` or `use_context` (respectively) argument that maps a list of context keys to the related config values:

**Example: Provide `test_value` as `example/context`**

```
[
	'name'              => 'example/provider',
    'config'            => [
        'test_value'        => [
            'type'              => 'text',
        ],
    ],
    'provides_context'  => [
        'example/context'   => 'test_value',
    ],
]
```

**Example: Use `example/context` value as the `test_value`**

```
[
	'name'              => 'example/consumer',
    'config'            => [
        'test_value'        => [
            'type'              => 'text',
        ],
    ],
    'use_context'       => [
        'example/context'   => 'test_value',
    ],
]
```

During template parsing, components get instantiated and hydrated first by defaults from their registered values (if they exists) and then context values are applied by passing the context store the the `Component::use_context()` and `Component::provides_context` methods.

Irving provides one default context value at runntime: the `irving/post` context, which is equal to the current global post ID.

### CODE REVIEW

File: inc/templates.php [hunk 1]
```diff
@@ -322,50 +323,72 @@ function convert_blocks_to_components( array $blocks ): array {
 }
 
 /**
- * Recursively iterate on a component tree.
+ * Hydrate components
```
Comment (jameswburke) [inline inc/templates.php:326]:
```suggestion
 * Hydrate components.
```

File: tests/test-context-store.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ *
+ * @group context
+ */
+class Test_Context_Store extends WP_UnitTestCase {
+
+	/**
+	 *
+	 * @var WP_Irving\Context_Store
+	 */
+	public static $context;
```
Comment (jameswburke) [inline tests/test-context-store.php:23]: Something missing?

File: inc/class-component.php [hunk 3]
```diff
@@ -652,24 +740,14 @@ public function jsonSerialize(): array {
 	 * @return array
 	 */
 	public function to_array(): array {
-
-		// Camel case config keys.
-		$this->set_config( $this->camel_case_keys( $this->get_config() ) );
-
-		// Camel case theme options.
-		$this->set_theme_options( $this->camel_case_keys( $this->get_theme_options() ) );
-		$this->set_theme( self::camel_case( $this->get_theme() ) );
```
[the remark this answers is not in the document]
Comment (joemcgill) [reply inc/class-component.php:661]: They get camel cased after they are added to the config, as part of the config camel casing.

File: inc/class-context-store.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	/**
+	 * Get the value of a context key.
+	 *
+	 * @param string $key The context key to retrieve.
+	 * @return mixed The value being returned
+	 */
+	public function get( string $key ) {
+		return isset( $this->get_context()[ $key ] ) ? $this->get_context()[ $key ] : null;
```
[the remark this answers is not in the document]
Comment (joemcgill) [reply inc/class-context-store.php:40]: Nope. Just forgot to use it here. Will update.

File: inc/class-component.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (jameswburke) [reply inc/class-component.php:661]: The config keys will, but we need to camelcase the values as well. I'm not seeing that happen anywhere?

**Pull request merged**